src/Facet/Driver.hs

```diff
@@ -90,46 +93,65 @@ kernel = Module kernelName [] [] $ Scope mempty
+      Nothing -> do
+        outputDocLn $ annotate Progress (brackets (ratio i nModules)) <+> nest 2 (group (fillSep [ pretty "Skipping", prettyMName name ]))
+        pure Nothing
   let nSuccess = length (catMaybes results)
       status
         | nModules == nSuccess = annotate Success (pretty nModules)
         | otherwise            = annotate Failure (ratio nSuccess nModules)
   outputDocLn (fillSep [status, reflow "modules loaded."])
   where
   ratio n d = pretty n <+> pretty "of" <+> pretty d
-  toNode (n, path, source, imports) = let imports' = map (Import.name . S.out) imports in Node n imports' (n, path, source, imports')
 
-loadModuleHeader :: (Has (Output :+: Throw (Notice.Notice (Doc Style))) sig m, MonadIO m) => [FilePath] -> Either FilePath MName -> m (MName, FilePath, Source, [S.Ann S.Import])
+data ModuleHeader a = ModuleHeader
+  { moduleName :: MName
+  , source     :: Source
+  , imports    :: [a]
+  }
+  deriving (Foldable, Functor, Traversable)
+
+imports_ :: Lens (ModuleHeader a) (ModuleHeader b) [a] [b]
+imports_ = lens imports (\ h imports -> h{ imports })
+
+headerNode :: ModuleHeader MName -> Node (ModuleHeader MName)
+headerNode h@(ModuleHeader n _ imports) = Node n imports h
+
+loadModuleHeader :: (Has (Output :+: Throw (Notice.Notice (Doc Style))) sig m, MonadIO m) => [FilePath] -> Either FilePath MName -> m (ModuleHeader MName)
 loadModuleHeader searchPaths target = do
   path <- case target of
     Left path  -> pure path
     Right name -> resolveName searchPaths name
   src <- rethrowIOErrors [] $ readSourceFromFile path
   -- FIXME: validate that the name matches
   (name', is) <- rethrowParseErrors @Style (runParserWithSource src (runFacet [] (whiteSpace *> moduleHeader)))
-  pure (name', path, src, is)
+  pure (ModuleHeader name' src (map (Import.name . S.out) is))
 
-loadModule :: Has (Output :+: State Options :+: State Target :+: Throw (Notice.Notice (Doc Style)) :+: Write (Notice.Notice (Doc Style))) sig m => MName -> FilePath -> Source -> [MName] -> m Module
-loadModule name path src imports = do
-  graph <- use modules_
-  let ops = foldMap (\ name -> lookupM name graph >>= map (\ (op, assoc) -> (name, op, assoc)) . operators . snd) imports
+loadModule :: Has (Output :+: State Options :+: Throw (Notice.Notice (Doc Style)) :+: Write (Notice.Notice (Doc Style))) sig m => Graph -> ModuleHeader Module -> m Module
+loadModule graph (ModuleHeader _ src imports) = do
+  let ops = foldMap (\ m -> map (\ (op, assoc) -> (name m, op, assoc)) (operators m)) imports
   m <- rethrowParseErrors @Style (runParserWithSource src (runFacet (map makeOperator ops) (whole module')))
   opts <- get
-  m <- rethrowElabWarnings . rethrowElabErrors opts . runReader graph . runReader src $ Elab.elabModule m
-  modules_.at name .= Just (Just path, m)
-  pure m
+  rethrowElabWarnings . rethrowElabErrors opts . runReader graph . runReader src $ Elab.elabModule m
```

Comment:
`loadModule` no longer reads from or writes to the `Target`, which makes it a little more composable.